pyteal/ast/abi/util_test.py

```diff
@@ -772,3 +772,250 @@ def test_sdk_type_specs_from_signature(sig_str, sig_args, sig_rets):
+        for safe_bidirectional in SAFE_BIDIRECTIONAL_TEST_CASES:
+            for t in safe_bidirectional.xs:
+                if type(t) == _ts:
+                    return True
+        return False
+
+    if isabstract(ts):
+        return
+    assert exists_in_safe_bidirectional(ts)
+
+
+class SafeAssignment(NamedTuple):
+    a: abi.TypeSpec
+    bs: list[abi.TypeSpec]
+
+
+SAFE_ASSIGNMENT_TEST_CASES: list[SafeAssignment] = [
+    SafeAssignment(
+        abi.StringTypeSpec(),
+        [abi.DynamicBytesTypeSpec(), abi.DynamicArrayTypeSpec(abi.ByteTypeSpec())],
+    ),
+    SafeAssignment(
+        abi.type_spec_from_annotation(NamedTDecl),
+        [
+            abi.type_spec_from_annotation(
+                abi.Tuple3[
+                    abi.Uint64,
+                    abi.Tuple3[
+                        abi.Transaction,
+                        abi.Address,
+                        abi.StaticArray[abi.Byte, Literal[16]],
+                    ],
+                    abi.Transaction,
+                ]
+            ),
+        ],
+    ),
+    SafeAssignment(
+        abi.type_spec_from_annotation(
+            abi.Tuple3[
+                abi.Uint64,
+                abi.Tuple3[
+                    abi.PaymentTransaction,
+                    abi.Address,
+                    abi.StaticArray[abi.Byte, Literal[16]],
+                ],
+                abi.PaymentTransaction,
+            ]
+        ),
+        [abi.type_spec_from_annotation(NamedTDecl)],
+    ),
+    SafeAssignment(
+        abi.AddressTypeSpec(),
+        [abi.StaticArrayTypeSpec(abi.ByteTypeSpec(), 32), abi.StaticBytesTypeSpec(32)],
+    ),
+    SafeAssignment(
+        abi.type_spec_from_annotation(NamedTComp0),
+        [abi.type_spec_from_annotation(NamedTComp1)],
+    ),
+] + [
+    SafeAssignment(spec, [abi.TransactionTypeSpec()])
+    for spec in abi.TransactionTypeSpecs
+    if spec != abi.TransactionTypeSpec()
+]
+
+
+@pytest.mark.parametrize("tc", SAFE_ASSIGNMENT_TEST_CASES)
+def test_type_spec_is_assignable_safe_assignment(tc: SafeAssignment):
+    assert len(tc.bs) > 0
+    for b in tc.bs:
+        assert abi.type_spec_is_assignable_to(tc.a, b)
+        assert not abi.type_spec_is_assignable_to(b, tc.a)
+
+
+@pytest.mark.parametrize("ts", bfs_on_inheritance(abi.TypeSpec))
+def test_type_spec_is_assignable_safe_assignment_full_coverage(ts: type):
+    def exists_in_safe_assignment(_ts: type):
+        for safe_assignment in SAFE_ASSIGNMENT_TEST_CASES:
+            if type(safe_assignment.a) == _ts:
+                return True
+            for t in safe_assignment.bs:
+                if type(t) == _ts:
+                    return True
+        return False
+
+    if isabstract(ts) or all(isabstract(parent) for parent in ts.__bases__):
+        return
+    assert exists_in_safe_assignment(ts)
+
+
+class UnsafeBidirectional(NamedTuple):
+    xs: list[abi.TypeSpec]
+
+
+UNSAFE_BIDIRECTIONAL_TEST_CASES: list[UnsafeBidirectional] = [
+    UnsafeBidirectional(
```

Comment:
@ahangsu Minor:  I think these uint cases can be added for more coverage.  Are you comfortable bringing them in?

```suggestion
    UnsafeBidirectional(
        [
            abi.Uint8TypeSpec(),
            abi.Uint16TypeSpec(),
            abi.Uint32TypeSpec(),
            abi.Uint64TypeSpec(),
        ]
    ),
    UnsafeBidirectional(
        [
            abi.ByteTypeSpec(),
            abi.Uint16TypeSpec(),
            abi.Uint32TypeSpec(),
            abi.Uint64TypeSpec(),
        ]
    ),
    UnsafeBidirectional(
```

Reply:
sure, that makes sense to me